parsec/core/fs/workspacefs/versioning_helpers.py

```diff
@@ -189,227 +309,245 @@ def __init__(self, remote_loader):
-        entry_id: EntryID,
-        early: Pendulum,
-        late: Pendulum,
-        version_number: int,
-        next_version_number: int,
-    ):
-        if early > late:
-            return
-        manifest = await manifest_cache.load(entry_id, version=version_number)
-        data = ManifestDataAndMutablePaths(
-            ManifestData(
-                manifest.author,
-                manifest.updated,
-                is_folder_manifest(manifest),
-                None if not is_file_manifest(manifest) else manifest.size,
+            task_list = VersionListerTaskList(
+                ManifestCacheCounter(self.manifest_cache, max_manifest_queries),
+                self.versions_list_cache,
             )
-        )
-        if len(target.parts) == path_level:
-
-            async def _populate_source_path(data, entry_id, timestamp):
-                data.source_path = await _try_get_path_at_timestamp(entry_id, timestamp)
-
-            async def _populate_destination_path(data, entry_id, timestamp):
-                data.destination_path = await _try_get_path_at_timestamp(entry_id, timestamp)
-
-            async def _populate_current_path(data, entry_id, timestamp):
-                data.current_path = await _try_get_path_at_timestamp(entry_id, timestamp)
-
-            # TODO : Use future manifest source field to follow files and directories
-            async with trio.open_service_nursery() as child_nursery:
-                child_nursery.start_soon(
-                    _populate_source_path, data, entry_id, early.add(microseconds=-1)
-                )
-                child_nursery.start_soon(_populate_destination_path, data, entry_id, late)
-                child_nursery.start_soon(_populate_current_path, data, entry_id, early)
-            tree[
-                TimestampBoundedEntry(manifest.id, manifest.version, early, late)
-            ] = ManifestDataAndPaths(
-                data=data.manifest,
-                source=data.source_path if data.source_path != data.current_path else None,
-                destination=data.destination_path
-                if data.destination_path != data.current_path
-                else None,
+            task_list.add(
+                starting_timestamp or root_manifest.created,
+                VersionListerTask(
+                    _populate_tree_list_versions,
+                    path,
+                    0,
+                    return_tree,
+                    root_manifest.id,
+                    starting_timestamp or root_manifest.created,
+                    ending_timestamp or Pendulum.now(),
+                ),
+            )
+            while not task_list.is_empty():
+                await task_list.execute_one()
+        except ManifestCacheDownloadLimitReached:  # TODO : what is limit?..
+            download_limit_reached = False
+        versions_list = [
+            TimestampBoundedData(*item[0], *item[1].data, item[1].source, item[1].destination)
```

Comment:
Ce serait mieux de faire `TimestampBoundedData(id=..., version=..., early=..., etc.)` car d'expérience on se retrouve très souvent avec un bug plus ou moins foireux à base d'inversion de paramètres suite à une refacto ou une inattention...

De même il faut mieux faire `for x, y in ...` plutôt que `item[0]` et `item[1]` pour éviter de croire que item est une sous-liste.

Autre question: pourquoi `*item[0], *item[1].data` ? Ça semble étrange vu que `item[0]` et `item[1].data`  sont dépendant l'un de l'autre ainsi que du nombre de paramètres que prend `TimestampBoundedData`...

Reply:
Le nombre de paramètre est fixe donc c'est safe, et j'avais juste évité de remettre en value ce qui servait déjà en clé. Mais finalement j'ai explicité les paramètres.

Reply:
Il y a trop de paramètres dans le constructeur de `TimestampBoundedData`, il faut mieux utiliser les passer sous la forme `key=value` (et pareil dans `_sanitize_list`)

Reply:
Done